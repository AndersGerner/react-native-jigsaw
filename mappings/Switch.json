{"name":"Switch","tag":"Switch","category":"input","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Control_Toggle.png","props":{"disabled":{"label":"Disabled","description":"Boolean to handle disabling the switch","required":false,"editable":true,"value":false,"type":"boolean"},"color":{"label":"Color","description":"Custom color for switch","editable":true,"value":null,"required":false,"type":"color"},"fieldName":{"label":"Field name","description":"The name of the field within the screen that will store this component's value","type":"fieldName","value":"switchValue","valuePropName":"value","handlerPropName":"onValueChange","editable":true,"required":false}},"layout":{}}