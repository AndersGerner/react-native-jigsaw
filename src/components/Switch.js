/* @flow */
import * as React from 'react';
import { Switch as NativeSwitch, Platform } from 'react-native';
import { withTheme } from '../core/theming';
import type { Theme } from '../types';
import { COMPONENT_TYPES, FORM_TYPES } from "../core/component-types";

type Props = {
  /**
   * Disable toggling the switch.
   */
  disabled?: boolean,
  /**
   * Value of the switch, true means 'on', false means 'off'.
   */
  value?: boolean,
  /**
   * Custom color for switch.
   */
  color?: string,
  /**
   * Callback called with the new value when it changes.
   */
  onValueChange?: Function,
  style?: any,
  /**
   * @optional
   */
  theme: Theme,
};

class Switch extends React.Component<Props> {
  render() {
    const {
      value,
      disabled,
      onValueChange,
      color,
      theme,
      ...props
    } = this.props;
    let thumbColor = undefined
    let checkedColor = color || theme.colors.primary
    if (Platform.OS !== 'ios') {
      thumbColor = theme.colors.surface
    }
    return (
      <NativeSwitch
        {...props}
        value={value}
        disabled={disabled}
        trackColor={{false: null, true: checkedColor }}
        thumbColor={thumbColor}
        onValueChange={disabled ? undefined : onValueChange}
        style={{ opacity: (disabled && Platform.OS !== 'ios') ? theme.disabledOpacity : 1 }}
      />
    );
  }
}

export const SEED_DATA = {
  name: "Switch",
  tag: "Switch",
  category: COMPONENT_TYPES.primitive,
  preview_image_url:
    "https://res.cloudinary.com/altos/image/upload/v1541096671/draftbit/library/jigsaw-1.0/reps/Control_Toggle.png",
  props: {
    disabled: {
      label: "Disable",
      description: "Boolean to handle disabling the switch",
      required: false,
      editable: true,
      value: false,
      type: FORM_TYPES.boolean,
    },
    value: {
      label: "status of switch controller",
      description: "boolean status if switch is toggled true or false",
      editable: true,
      value: true,
      type: FORM_TYPES.boolean,
    },
    color: {
      label: "Switch color",
      description: "Custom color for switch",
      editable: true,
      value: null,
      required: false,
      type: FORM_TYPES.color,
    },
    onValueChange: {
      label: "Switch onPress handler",
      description: "Value of the switch, true means 'on', false means 'off'",
      editable: true,
      required: true,
      value: "{this.onPress}",
      type: FORM_TYPES.function,
    },
  },
  layout: {
    width: 36,
    height: 22
  },
};

export default withTheme(Switch);