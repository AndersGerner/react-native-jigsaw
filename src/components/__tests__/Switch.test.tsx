import * as React from "react";
import renderer from "react-test-renderer";
import Switch from "../Switch";

it("renders switch", () => {
  const tree = renderer.create(<Switch value={false} />).toJSON();

  expect(tree).toMatchSnapshot();
});
