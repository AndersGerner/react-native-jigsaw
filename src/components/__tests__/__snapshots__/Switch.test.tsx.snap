// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders switch 1`] = `
<RCTSwitch
  accessibilityRole="switch"
  activeTrackColor="rgba(90, 69, 255, 1)"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onTintColor="rgba(90, 69, 255, 1)"
  style={
    Array [
      Object {
        "height": 31,
        "width": 51,
      },
      Array [
        Object {
          "opacity": 1,
        },
        undefined,
      ],
    ]
  }
  tintColor=""
  value={false}
/>
`;
